File: Kotlin/sort/insertion_sort.kt
fun insertion_sort(A: Array<Int>) {
    var n = A.size
    var i: Int
    for (j in 1 until n) {
        var key = A[j]
        i = j - 1
        while (i >= 0 && A[i] > key) {
            A[i + 1] = A[i]
            i--
        }
        A[i + 1] = key
    }
}

fun main(arg: Array<String>) {
    print("Enter no. of elements :")
    var n = readLine()!!.toInt()

    println("Enter elements : ")
    var A = Array(n, { 0 })
    for (i in 0 until n)
        A[i] = readLine()!!.toInt()

    insertion_sort(A)

    println("Sorted array is : ")
    for (i in 0 until n)
        print("${A[i]}  ")
}

/*
Output:

Enter no. of elements: 5
Enter elements:
22
12
33
32
11
Sorted array is:
11  12  22  32  33

*/

File: PHP/sort/InsertionSort.php
<?php
function InsertionSort(&$arr)
{
    $n=count($arr);
    $next=null;
    for($i=1; $i<$n; $i++)//outer loop
    {
        $next=$arr[$i];
        for($j=$i-1; $j>=0; $j--)//inner loop
        {
            if( $arr[$j]>$next )//change > to < for descending order
            {
                $arr[$j+1]=$arr[$j];
            }
            else
            {
                break;
            }
        }
        $arr[$j+1]=$next; // insert the next value to the correct postion of the already sorted elements
 
    }
 
    return $arr;
}


$arr = array(-1, 2, -8, 854, -1900, 0, 250);
echo "Array before sorting:"."\n";
for($i=0;$i<count($arr);$i++)
{
    echo $arr[$i]." ";
}
InsertionSort($arr);
echo "\nArray after sorting:"."\n";
for($i=0;$i<count($arr);$i++)
{
    echo $arr[$i]." ";
}
/*
Sample Output
Array before sorting:
-1 2 -8 854 -1900 0 250 
Array after sorting:
-1900 -8 -1 0 2 250 854 

Time Complexity

    Best complexity: n
    Average complexity: n^2
    Worst complexity: n^2

*/


?>
